File: packages/swift/Sources/CatalystSwift/Models/CatalystFleet.swift
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

import Foundation

public struct CatalystFleetPlacement: Decodable, Sendable, Equatable, Hashable {
  public let posX: Double
  public let posY: Double
  public let scale: Double
  public let rotation: Double
}

public struct CatalystFleetText: Decodable, Sendable, Equatable, Hashable {
  public let id: String
  public let body: String
  public let textStyle: String
  public let textAlignment: String
  public let color: String
  public let backgroundColor: String?
  public let posX: Double
  public let posY: Double
  public let scale: Double
  public let rotation: Double
}

public struct CatalystFleetSticker: Decodable, Sendable, Equatable, Hashable {
  public let id: String
  public let emoji: String
  public let posX: Double
  public let posY: Double
  public let scale: Double
  public let rotation: Double
}

public struct CatalystFleetMedia: Decodable, Sendable, Equatable, Hashable {
  public let url: String
  public let alt: String
  public let width: Int?
  public let height: Int?
  public let placement: CatalystFleetPlacement?
}

public struct CatalystFleet: Decodable, Sendable, Equatable, Hashable {
  public let id: String
  public let backgroundColor: String
  public let renderedImageUrl: String?
  public let user: EgeriaUser
  public let texts: [CatalystFleetText]
  public let media: CatalystFleetMedia?
  public let stickers: [CatalystFleetSticker]
  public let reactions: [String: CatalystReaction]
  public let viewCount: Int
  public let createdAt: Date
  public let expiresAt: Date
}

public struct CatalystFleetViewer: Decodable, Sendable, Equatable, Hashable {
  public let user: EgeriaUser
  public let viewedAt: Date
}

public struct CatalystFleetRing: Decodable, Sendable, Equatable, Hashable {
  public let user: EgeriaUser
  public let hasUnread: Bool
  public let fleetCount: Int
}

File: packages/kotlin/src/main/kotlin/com/natsuneko/catalyst/models/CatalystFleet.kt
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

package com.natsuneko.catalyst.models

import kotlinx.serialization.Serializable
import kotlin.time.ExperimentalTime

/**
 * Placement (position/scale/rotation) of a fleet element
 */
@Serializable
data class CatalystFleetPlacement(
    val posX: Double,
    val posY: Double,
    val scale: Double,
    val rotation: Double
)

/**
 * A text overlay on a fleet
 */
@Serializable
data class CatalystFleetText(
    val id: String,
    val body: String,
    val textStyle: String,
    val textAlignment: String,
    val color: String,
    val backgroundColor: String? = null,
    val posX: Double,
    val posY: Double,
    val scale: Double,
    val rotation: Double
)

/**
 * A sticker overlay on a fleet
 */
@Serializable
data class CatalystFleetSticker(
    val id: String,
    val emoji: String,
    val posX: Double,
    val posY: Double,
    val scale: Double,
    val rotation: Double
)

/**
 * Media attached to a fleet
 */
@Serializable
data class CatalystFleetMedia(
    val url: String,
    val alt: String,
    val width: Int? = null,
    val height: Int? = null,
    val placement: CatalystFleetPlacement? = null
)

/**
 * A fleet (ephemeral status)
 */
@Serializable
data class CatalystFleet @OptIn(ExperimentalTime::class) constructor(
    val id: String,
    val backgroundColor: String,
    val renderedImageUrl: String? = null,
    val user: EgeriaUser,
    val texts: List<CatalystFleetText>,
    val media: CatalystFleetMedia?,
    val stickers: List<CatalystFleetSticker>,
    val reactions: Map<String, CatalystReaction>,
    val viewCount: Int,
    val createdAt: kotlin.time.Instant,
    val expiresAt: kotlin.time.Instant
)

/**
 * A viewer of a fleet
 */
@Serializable
data class CatalystFleetViewer @OptIn(ExperimentalTime::class) constructor(
    val user: EgeriaUser,
    val viewedAt: kotlin.time.Instant
)

/**
 * A user's fleet ring summary (used for the fleet ring/tray UI)
 */
@Serializable
data class CatalystFleetRing(
    val user: EgeriaUser,
    val hasUnread: Boolean,
    val fleetCount: Int
)
